pkg/modules/notary/notary.go

```diff
@@ -0,0 +1,300 @@
+	"github.com/nspcc-dev/neo-go/pkg/core/blockchainer"
+	"github.com/nspcc-dev/neo-go/pkg/core/mempool"
+	"github.com/nspcc-dev/neo-go/pkg/core/transaction"
+	"github.com/nspcc-dev/neo-go/pkg/io"
+	"github.com/nspcc-dev/neo-go/pkg/network/payload"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+	"github.com/nspcc-dev/neo-go/pkg/vm"
+	"github.com/nspcc-dev/neo-go/pkg/vm/emit"
+	"github.com/nspcc-dev/neo-go/pkg/wallet"
+	"go.uber.org/zap"
+)
+
+// Notary represents Notary module.
+type (
+	Notary struct {
+		Config Config
+
+		// mtx protects setting callbacks.
+		mtx sync.RWMutex
+		// onTransaction is a callback for completed transactions (mains or fallbacks) sending.
+		onTransaction func(tx *transaction.Transaction) error
+
+		// reqMtx protects requests list.
+		reqMtx sync.RWMutex
+		// requests represents the map of main transactions which needs to be completed
+		// with the associated fallback transactions grouped by the main transaction hash
+		requests map[util.Uint256]request
+
+		// accMtx protects account.
+		accMtx      sync.RWMutex
+		currAccount *wallet.Account
+		wallet      *wallet.Wallet
+	}
+
+	Config struct {
+		MainCfg config.P2PNotary
+		Chain   blockchainer.Blockchainer
+		Log     *zap.Logger
+	}
+)
+
+// request represents Notary service request.
+type request struct {
+	typ RequestType
+	// isValid indicates whether main transaction satisfies Notary node verification rules and can be sent to the network.
+	isValid bool
+	// isSent indicates whether main transaction was successfully sent to the network.
+	isSent bool
+	main   *transaction.Transaction
+	// minNotValidBefore is the minimum NVB value among fallbacks transactions.
+	// We stop trying to send mainTx to the network if the chain reaches minNotValidBefore height.
+	minNotValidBefore uint32
+	fallbacks         []*transaction.Transaction
+	// nKeys is the number of witnesses to be collected for the main transaction
+	nKeys uint8
+	// nKeysCollected is the number of already collected witnesses
+	nKeysCollected uint8
+}
+
+// NewNotary returns new Notary module.
+func NewNotary(bc blockchainer.Blockchainer, log *zap.Logger) (*Notary, error) {
+	cfg := bc.GetConfig().Modules.P2PNotary
+	w := cfg.UnlockWallet
+	wallet, err := wallet.NewWalletFromFile(w.Path)
+	if err != nil {
+		return nil, err
+	}
+
+	haveAccount := false
+	for _, acc := range wallet.Accounts {
+		if err := acc.Decrypt(w.Password); err == nil {
+			haveAccount = true
+			break
+		}
+	}
+	if !haveAccount {
+		return nil, errors.New("no wallet account could be unlocked")
+	}
+
+	return &Notary{
+		requests: make(map[util.Uint256]request),
+		Config: Config{
+			MainCfg: cfg,
+			Chain:   bc,
+			Log:     log,
+		},
+		wallet: wallet,
+	}, nil
+}
+
+// OnNewRequest is a callback method which is called after new notary request is added to the notary request pool.
+func (n *Notary) OnNewRequest(payload *payload.P2PNotaryRequest) {
+	if n.getAccount() == nil {
+		return
+	}
+
+	n.reqMtx.Lock()
```

Comment:
We can reduce this critical section a little, you only need it at line 115.

Reply:
Fixed